File: scripts/jeremy-post-install.go
// jeremy-post-install
//
// This is my hook to run arbitrary code after installing my nix packages.
//
// It is very sad that I need this hook.
//
// Strive to keep this as small as possible!
//
// Also, everything in here must be idempotent and main should complete in
// under ~50ms.

package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

func main() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Fprintf(os.Stderr, "error during jeremy-post-install: %v\n", r)
		}
	}()

	nixConfDir := os.ExpandEnv("$HOME/.nix-profile/config/")
	homeConfDir := os.ExpandEnv("$HOME/.config/")

	// Make ~/.config if it does not yet exist.
	os.MkdirAll(homeConfDir, 0755)

	wantSymlink := make(map[string]bool)

	// Symlink ~/.nix-profile/config/* into ~/.config
	{
		des, err := os.ReadDir(nixConfDir)
		check(err)
		for _, de := range des {
			if de.Name() == "README.md" {
				continue
			}
			wantSymlink[de.Name()] = true
			linkFrom := homeConfDir + de.Name()
			haveLink, err := os.Readlink(linkFrom)
			wantLink := nixConfDir + de.Name()
			switch {
			case os.IsNotExist(err):
				fmt.Printf("symlinking %v config...\n", de.Name())
				check(os.Symlink(wantLink, linkFrom))
			case err == nil && haveLink == wantLink:
				// Nothing more to do.
			case err == nil && haveLink != wantLink:
				fmt.Printf("I want to install a symlink at %v, but it is already symlinked to %q\n", linkFrom, haveLink)
			default:
				_, err = os.Stat(linkFrom)
				if err == nil {
					fmt.Printf("I want to install a symlink at %v, but there is already something else there.\n", linkFrom)
				} else {
					fmt.Printf("I want to install a symlink at %v, but I failed to stat the existing file: %v\n", linkFrom, err)
				}
			}
		}
	}

	// Remove old symlinks from ~/.config
	{
		des, err := os.ReadDir(homeConfDir)
		check(err)
		for _, de := range des {
			link, err := os.Readlink(homeConfDir + de.Name())
			if err == nil && strings.HasPrefix(link, nixConfDir) && !wantSymlink[de.Name()] {
				fmt.Printf("removing %v config...\n", de.Name())
				check(os.Remove(homeConfDir + de.Name()))
			}
		}
	}

	// Build bat theme
	{
		if _, err := os.Stat(filepath.Join(os.Getenv("HOME"), ".cache", "bat", "themes.bin")); os.IsNotExist(err) {
			fmt.Println("installing solarized theme for bat...")
			run(exec.Command("bat", "cache", "--build"))
		}
	}
}

func run(cmd *exec.Cmd) {
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	check(cmd.Run())
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

File: scripts/jeremy-post-install.rs
extern crate dirs;

use anyhow::{anyhow, bail, Result};
use std::{collections, fs, io::ErrorKind, os, process};

fn main() -> Result<()> {
    let home = dirs::home_dir().ok_or(anyhow!("no HOME"))?;
    let nix_conf = home.join(".nix-profile/config");
    let home_conf = home.join(".config");

    // Make ~/.config if it does not yet exist.
    fs::create_dir_all(&home_conf)?;

    let mut want_symlink = collections::HashSet::new();

    // Symlink ~/.nix-profile/config/* into ~/.config
    for entry in fs::read_dir(&nix_conf)? {
        let name = entry?.file_name();
        if name == "README.md" {
            continue;
        }
        want_symlink.insert(name.clone());
        let link_from = home_conf.join(&name);
        let have_link = fs::read_link(&link_from);
        let want_link = nix_conf.join(&name);
        match have_link {
            Ok(p) if p == want_link => (), // Nothing more to do.
            Ok(p) => println!(
                "I want to install a symlink at {:?}, but it is already symlinked to {:?}",
                link_from, p
            ),
            Err(err) if err.kind() == ErrorKind::NotFound => {
                println!("symlinking {:?} config...", name);
                os::unix::fs::symlink(want_link, link_from)?;
            }
            Err(_) => match fs::metadata(&link_from) {
                Ok(_) => println!(
                    "I want to install a symlink at {:?}, but there is already something else there.",
                    link_from
                ),
                Err(err) => println!(
                    "I want to install a symlink at {:?}, but I failed to stat the existing file: {}",
                    link_from, err
                ),
            },
        }
    }

    // Remove old symlinks from ~/.config
    for entry in fs::read_dir(&home_conf)? {
        let entry = entry?;
        let name = entry.file_name();
        match fs::read_link(home_conf.join(&name)) {
            Ok(link) if link.starts_with(&nix_conf) && !want_symlink.contains(&name) => {
                println!("removing {:?} config...", name);
                fs::remove_file(entry.path())?;
            }
            _ => (), // doesn't matter
        }
    }

    // Build bat theme
    match fs::metadata(home.join(".cache/bat/themes.bin")) {
        Err(err) if err.kind() == ErrorKind::NotFound => {
            println!("installing solarized theme for bat...");
            if !process::Command::new("bat")
                .args(["cache", "--build"])
                .status()?
                .success()
            {
                bail!("`bat cache --build` failed");
            }
            Ok(())
        }
        _ => Ok(()),
    }
}
